### PR TITLE
chore: `EventManager` add `cancelEvent` method

### LINKED ISSUE #14451: [CAL-3345] Add `cancel` method to `EventManager`
<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-3345](https://linear.app/calcom/issue/CAL-3345/add-cancel-method-to-eventmanager)</sub>

### PR DESCRIPTION
**Pull request by joeauyeung**
## What does this PR do?

This PR adds the `cancelEvent` method to the `EventManager` class. Logic surrounding deleting 3rd party data when canceling a booking lived in both `handleCancelBooking` and the `EventManager` in a private class. 

My hope with this PR is to refactor this so this logic only lives inside of the `EventManager`

Fixes #14451 

## Requirement/Documentation

- If there is a requirement document, please, share it here.
- If there is a UI/UX design document, please, share it here.

## Type of change

- Chore (refactoring code, technical debt, workflow improvements)

## How should this be tested?

- Install a calendar and a video integration and enable them on an event type
	- I used GCal and Zoom
- Create a booking
	- The references should be created in the 3rd party apps
- Delete the booking
	- The references should be deleted in the 3rd party apps	

## Mandatory Tasks

- [x] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

## Checklist

- I haven't added tests that prove my fix is effective or that my feature works
- I haven't checked if new and existing unit tests pass locally with my changes

### CODE REVIEW

File: packages/core/EventManager.ts [hunk 1]
```diff
@@ -246,25 +246,36 @@ export default class EventManager {
...
       credentialId,
       type: credentialType,
-    } = bookingCalendarReference;
+    } = reference;
 
-    const calendarCredential = await this.getCredentialAndWarnIfNotFound(credentialId, credentialType);
+    const bookingRefUid =
+      isBookingInRecurringSeries && reference?.thirdPartyRecurringEventId
+        ? reference.thirdPartyRecurringEventId
+        : reference.uid;
```
Comment (joeauyeung) [inline packages/core/EventManager.ts:272]: Check if we're deleting a booking in a recurring series. This change was made in `handleCancelBooking` but missing in the `EventManager`

File: packages/core/EventManager.ts [hunk 2]
```diff
@@ -275,50 +286,56 @@ export default class EventManager {
...
+      return credential;
+    } else {
+      const credential =
+        typeof credentialId === "number" && credentialId > 0
+          ? await prisma.credential.findUnique({
+              where: {
+                id: credentialId,
+              },
+              select: credentialForCalendarServiceSelect,
+            })
```
Comment (joeauyeung) [inline packages/core/EventManager.ts:320]: Before this function would query the database for the credential. When we create a new instance of the `EventManager` class we need to pass the credentials anyways. Let's search through the appropriate credential arrays first before querying the database.

File: packages/core/EventManager.ts [hunk 3]
```diff
@@ -451,30 +468,62 @@ export default class EventManager {
...
+      };
+    }>,
+    isBookingInRecurringSeries?: boolean
+  ) {
+    await this.deleteEventsAndMeetings({
+      event,
+      bookingReferences,
+      isBookingInRecurringSeries,
+    });
+  }
```
Comment (joeauyeung) [inline packages/core/EventManager.ts:489]: Expose the `deleteEventsAndMeetings` method.

File: packages/core/EventManager.ts [hunk 4]
```diff
@@ -451,30 +468,62 @@ export default class EventManager {
...
   }) {
-    const calendarReferences = booking.references.filter((reference) => reference.type.includes("_calendar"));
-    const videoReferences = booking.references.filter((reference) => reference.type.includes("_video"));
-    log.debug("deleteEventsAndMeetings", safeStringify({ calendarReferences, videoReferences }));
-    const calendarPromises = calendarReferences.map(async (bookingCalendarReference) => {
-      return this.deleteCalendarEventForBookingReference({
-        bookingCalendarReference,
-        event,
-      });
-    });
```
Comment (joeauyeung) [inline packages/core/EventManager.ts:469]: We're looping through the booking references two times.

File: packages/core/EventManager.ts [hunk 5]
```diff
@@ -451,30 +468,62 @@ export default class EventManager {
...
+    const calendarReferences = [],
+      videoReferences = [],
+      allPromises = [];
+
+    for (const reference of bookingReferences) {
+      if (reference.type.includes("_calendar")) {
+        calendarReferences.push(reference);
+        allPromises.push(
+          this.deleteCalendarEventForBookingReference({
```
Comment (joeauyeung) [inline packages/core/EventManager.ts:508]: Let's loop through the references once and call the appropriate method.

File: packages/core/EventManager.ts [hunk 6]
```diff
@@ -451,30 +468,62 @@ export default class EventManager {
...
+        allPromises.push(
+          this.deleteVideoEventForBookingReference({
+            reference,
+          })
+        );
+      }
+    }
 
-    const allPromises = [...calendarPromises, ...videoPromises];
+    log.debug("deleteEventsAndMeetings", safeStringify({ calendarReferences, videoReferences }));
```
Comment (joeauyeung) [inline packages/core/EventManager.ts:526]: I still kept this for logging though.

File: packages/features/bookings/lib/handleCancelBooking.ts
Comment (joeauyeung) [inline packages/features/bookings/lib/handleCancelBooking.ts:1]: With this refactor we're reducing the size of `handleCancelBooking` by 21%.

**Review by zomars — APPROVED** [review]
LGTM 

Nice work @joeauyeung 🙏🏽

**Review by zomars — CHANGES_REQUESTED** [review]
We're having some type errors in CI. Let's get this fixed before merging

File: packages/features/bookings/lib/getAllCredentialsForUsersOnEvent/getAllCredentials.ts [hunk 7]
```diff
@@ -14,13 +13,13 @@ type User = Prisma.UserGetPayload<typeof userSelect>;
  *
  */
 export const getAllCredentials = async (
-  user: User & { credentials: CredentialPayload[] },
-  eventType: Awaited<ReturnType<typeof getEventTypesFromDB>>
+  user: { id: number; username: string | null; credentials: CredentialPayload[] },
+  eventType: { team: { id: number | null } | null; parentId: number | null } | null
```
Comment (joeauyeung) [inline packages/features/bookings/lib/getAllCredentialsForUsersOnEvent/getAllCredentials.ts:17]: @zomars made some changes here. I think the types casted too broad of an umbrella. I narrowed it down to what's really needed. Thoughts?

**Review by sean-brydon — APPROVED** [review]

**Review by zomars — APPROVED** [review]
LGTM. Let's wait for the missing check

File: packages/features/bookings/lib/handleCancelBooking.ts [hunk 8]
```diff
@@ -553,13 +461,6 @@ async function handler(req: CustomRequest) {
   const prismaPromises: Promise<unknown>[] = [bookingReferenceDeletes];
```
Comment (zomars) [inline packages/features/bookings/lib/handleCancelBooking.ts:461]: This is now orphaned @joeauyeung is should we remove this?

**Pull request merged**